include/deal.II/lac/block_matrix_base.h

```diff
@@ -195,10 +195,13 @@ namespace BlockMatrixIterators
     operator==(const Accessor &a) const;
 
     template <typename>
-    friend class MatrixIterator;
+    friend class dealii::MatrixIterator;
```

Comment:
This is the context where it matters. The `friend` declaration references the tentative declaration previously created (in the current namespace), rather than the class we really wanted (in the global namespace). No such class is ever declared, so the tentative declaration will remain tentative forever, but the important point is that the class we wanted to be a friend will not actually be a friend.

Reply:
Interestingly, in line 284 below (for the `const` accessor case), we got this right and always had `friend class dealii::MatrixIterator`, with the namespace qualification.